fed/mosaic-application/src/main/java/org/eclipse/mosaic/fed/application/ambassador/simulation/perception/errormodels/SimpleOcclusionModifier.java

```diff
@@ -0,0 +1,122 @@
+
+import org.eclipse.mosaic.fed.application.ambassador.simulation.perception.PerceptionModuleOwner;
+import org.eclipse.mosaic.fed.application.ambassador.simulation.perception.SimplePerceptionConfiguration;
+import org.eclipse.mosaic.fed.application.ambassador.simulation.perception.VehicleObject;
+import org.eclipse.mosaic.lib.math.Vector3d;
+
+import org.apache.commons.lang3.Validate;
+
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+
+/**
+ * This {@link PerceptionModifier} tries to emulate the behaviour of occlusion.
+ * The general approach is to look at the angle of all vehicles that are
+ * perceived closer to the current vehicle.
+ * If any of those previously perceived vehicles has an angle smaller than
+ * defined by {@link #getOcclusionAngle} the current vehicle will not be perceived.
+ * <p/>
+ * Additionally, a linear function between {@link #minDetectionAngle} and {@link #maxDetectionAngle}
+ * is fitted, which makes it necessary for further vehicles to have a larger "free" angle.
+ */
+public class SimpleOcclusionModifier implements PerceptionModifier {
+
+    /**
+     * The "free" angle required by the closest vehicle. [rad]
+     */
+    private final double minDetectionAngle;
+
+    /**
+     * The "free" angle required by the furthest vehicle. [rad]
+     */
+    private final double maxDetectionAngle;
+
+    /**
+     * Constructor for the {@link SimpleOcclusionModifier}.
+     *
+     * @param minDetectionAngle the "free" angle that will be required by closest vehicles [degree]
+     * @param maxDetectionAngle the "free" angle that will be required by furthest vehicles [degree]
+     */
+    public SimpleOcclusionModifier(double minDetectionAngle, double maxDetectionAngle) {
+        Validate.isTrue(minDetectionAngle > 0 && maxDetectionAngle >= minDetectionAngle,
+                "Angles have to be larger than 0 and maxDetectionAngle >= minDetectionAngle.");
+        this.minDetectionAngle = Math.toRadians(minDetectionAngle);
+        this.maxDetectionAngle = Math.toRadians(maxDetectionAngle);
+    }
+
+    @Override
+    public List<VehicleObject> apply(PerceptionModuleOwner owner, List<VehicleObject> vehicleObjects) {
+        if (vehicleObjects.size() == 0) {
+            return vehicleObjects;
+        }
+        Vector3d ownerPosition = owner.getVehicleData().getProjectedPosition().toVector3d();
+        // sort by distances
+        List<VehicleObject> sortedByDistance = new ArrayList<>(vehicleObjects);
+        sortedByDistance.sort(Comparator.comparingDouble(vehicleObject -> vehicleObject.distanceTo(ownerPosition)));
+        // fit linear function to (closest distance, min angle) and (furthest distance, max angle)
+        double closestPerceivedDistance = ownerPosition.distanceTo(sortedByDistance.get(0));
+        double furthestPerceivedDistance;
+        if (owner.getPerceptionModule().getConfiguration().getClass().equals(SimplePerceptionConfiguration.class)) {
```

Comment:
see comment above. maybe we should add `getViewingRange()` to PerceptionConfiguration interface to avoid casting?

Reply:
Yes, I considered this as well. I suppose every type of perception will have some kind of range defined 👍